### PR TITLE
Configure preview workflow

### PR DESCRIPTION
**Pull request by minkimcello**
## Motivation

To setup preview-package-publishing workflow. I was asked to setup both preview and release but it looks like the release-publishing workflow has already been setup.

Published [here](https://www.npmjs.com/package/material-ui-interactors) using [this](https://github.com/thefrontside/material-ui-interactors/blob/v4/.github/workflows/covector-version-or-publish.yml) workflow.

## Approach

- New workflow that runs "preview" command
- setup-node step is included in workflow and double checked for `NPM_TOKEN` secret in this repository
- the default label to trigger the preview build is set as `preview`
  - attached the label on this PR to see if it works

### CODE REVIEW

**Comment by frontsidejack** [comment]
The following preview packages have been published by Covector:
- `material-ui-interactors@4.0.0-mk-publish.1626208194558`

<p align="right">b42579a</p>

File: .changes/config.json [hunk 1]
```diff
@@ -4,8 +4,7 @@
     "javascript": {
       "version": true,
       "getPublishedVersion": "npm view ${ pkgFile.pkg.name } version",
-      "publish": "npm publish --access public",
-      "errorOnVersionRange": "4.0.0-alpha.100 - 99.x || ^4.4.0-0 || ^5.0.0-0"
```
Comment (minkimcello) [inline .changes/config.json:8]: @cowboyd What was the reason for wanting to prevent those specific version ranges?

Comment (cowboyd) [reply .changes/config.json:8]: Not sure... @jbolda ?

**Comment by cowboyd** [comment]
>The following preview packages have been published by Covector:
>
> material-ui-interactors@4.0.0-alpha.1-mk-publish.1624289984022

This is pretty amazing! Is there a way to make the version number simpler?

like:

`material-ui-interactors@4.0.0-mk-publish.1`

The `mk-publish` should _replace_ the `alpha` tag since they are both pre-release version tags, and so it doesn't make sense to have two.

**Comment by cowboyd** [comment]
@minkimcello @jbolda any thoughts here?

**Comment by minkimcello** [comment]
> @minkimcello @jbolda any thoughts here?

@cowboyd oh i forgot to add a thumbs up to your first response. i think removing the pre-release identifier for the preview package versions makes sense. i'll put together a PR for it today

**Comment by cowboyd** [comment]
@minkimcello Is this ready for another shot?

**Comment by minkimcello** [comment]
> @minkimcello Is this ready for another shot?

I think we can go ahead and merge this in. I was waiting on what we should do about the semver-exceptions but as per my response in our Discord discussion, I don't think it's necessary.

**Comment by cowboyd** [comment]
Cool. Can you link/paste that discussion in here so we have a record?

**Comment by minkimcello** [comment]
https://ptb.discord.com/channels/700803887132704931/719594558853218395/859440451295510548

<blockquote><div> Discord</div><div><strong><a href="https://ptb.discord.com/channels/700803887132704931/719594558853218395/859440451295510548">Discord - A New Way to Chat with Friends & Communities</a></strong></div><div>Discord is the easiest way to communicate over voice, video, and text.  Chat, hang out, and stay close with your friends and communities.</div></blockquote>

**Comment by cowboyd** [comment]
@minkimcello also, before we merge, can we see what the preview that it will generate looks like?

**Comment by minkimcello** [comment]
> @minkimcello also, before we merge, can we see what the preview that it will generate looks like?

☝️ 

Screen Shot 2021-07-12 at 12 45 39 PM

**Comment by cowboyd** [comment]
Is there a way to have a simple version number like `material-ui-interactors@4.0.0-mk-publish.1`?

**Comment by minkimcello** [comment]
> Is there a way to have a simple version number like `material-ui-interactors@4.0.0-mk-publish.1`?

If I remember correctly, the reason why we don't do increments for the preview packages is that there's no simple way of figuring out what number should be next (as opposed to the normal pre-release process where we bump and merge via version PRs).

There are two identifiers available for the preview package versioning at the moment: date and sha.

**Comment by cowboyd** [comment]
Can we check with the NPM tag?

**Comment by minkimcello** [comment]
> Can we check with the NPM tag?

That was one approach we were considering; running `npm view package` and grabbing the latest version that way. But sometimes there are delays to the update and could cause problems if someone pushes commits too frequently. So with the delay and the manual parsing, it didn't seem like a robust solution.

**Comment by cowboyd** [comment]
> > Can we check with the NPM tag?
> 
> That was one approach we were considering; running `npm view package` and grabbing the latest version that way. But sometimes there are delays to the update and could cause problems if someone pushes commits too frequently. 

What kind of delays are we talking about, as in how often do we expect this to fail? I know that publishing to the `latest` tag for example has always felt pretty instantaneous.

> So with the delay and the manual parsing, it didn't seem like a robust solution.

I think we need to quantify robust here because the tradeoff that we'd be settling for would be a suboptimal DX.

**Comment by minkimcello** [comment]
> What kind of delays are we talking about, as in how often do we expect this to fail? I know that publishing to the `latest` tag for example has always felt pretty instantaneous.

It's often instantaneous but sometimes it takes a few minutes for the registry to update. This is likely not something that'll become an actual problem but it is something I noticed from all the experimental publishing I did for covector and github actions.

> I think we need to quantify robust here because the tradeoff that we'd be settling for would be a suboptimal DX.

So ideally it would be nice to do `^1.1.1-feature.0` so that it'll pull the latest preview package. But with these prerelease versions, if `1.1.1` already exists, `^1.1.1-feature.0` will install `1.1.1`.

One way around this is to patch bump to `1.1.2-feature.0` but then if `1.1.2` is published because someone else merges a `version packages` pr, the person working on the feature pr will start installing the new `1.1.2` or `1.2.0` instead of their preview package.

Also, another thing I noticed was there was some caching(?) issues with fetching some of these packages. Say if I have `^.0` and recently published `.1`, I should be able to install `^.0` and get `.1`. And afterwards if I publish `.2` and `.3`, installing `^.0` should get me `.3` but sometimes it will keep installing `.1` until I change my dependency to `^.2` which only _then_ it'll install `.3` and _then_ I can bump it back down to `^.0` to get `.3`.

It was at this point that we thought maybe it's good to be explicit in which preview package we want to install (instead of using carets). Also, it was my opinion that connecting preview packages to the commit sha is useful because the developer can check via `git log` and `npm view` to see if their preview package has been published. If i'm at `1.1.1-feature.8`, i might forget if `.8` was from an earlier commit or if it's the latest one.

[This](https://github.com/jbolda/covector/issues/147#issuecomment-812678832) is where I jot down my findings around prerelease versioning for covector (under the `Consuming Previews` section).

**Comment by minkimcello** [comment]
@jbolda Thoughts? ☝️ It'd be pretty easy to implement on covector.

**Comment by cowboyd** [comment]
@minkimcello This is very thoroughly researched. Awesome.

> So ideally it would be nice to do ^1.1.1-feature.0 so that it'll pull the latest preview package. But with these prerelease versions, if 1.1.1 already exists, ^1.1.1-feature.0 will install 1.1.1.

As a consumer of the package, if you're using a caret dependency, then that's what you asked for right? If it isn't, then you'll want to pin.

As an author of the package, if `1.1.1` is released, then there is no way for you to get your change in, so you'll have to rebase anyway.

>One way around this is to patch bump to 1.1.2-feature.0 but then if 1.1.2 is published because someone else merges a version packages pr, the person working on the feature pr will start installing the new 1.1.2 or 1.2.0 instead of their preview package.

Again, I don't think that you should be doing development with caret versions. It's only really reasonable for downstream consumers to use carets, and if a compatible version is released, then that's probably going to be better for you (which won't even happen until you update your `yarn.lock`). Even if you were developing with caret versions, a release ahead of you means that your changes were never going to make it out the door with that release, and so you're going to have to rebase onto the next one.

Still. It makes me think that maybe having clean numbers isn't the thing that will provide a good DX, but instead clean labels?

In other words, I just want to say `material-ui-interactors@mk--publish` and get whatever version it is that currently corresponds to that label. This is, in effect what you _always_ do when you install a package without a version, is that you're fetching the `latest` label.

**Comment by jbolda** [comment]
> It's often instantaneous but sometimes it takes a few minutes for the registry to update. This is likely not something that'll become an actual problem but it is something I noticed from all the experimental publishing I did for covector and github actions.

Luckily with effection this shouldn't be a huge issue to deal with in covector. This would be opt-in for the user, so they would be subscribing to the possibility that npm has an issue. If they are pulling a version and it fails, it is likely that it would have the same issue on publish. We needed the current functionality and structure regardless, but now we could look at layer `npm view ...` on top of it if we choose.

> > I think we need to quantify robust here because the tradeoff that we'd be settling for would be a suboptimal DX.
> 
> So ideally it would be nice to do `^1.1.1-feature.0` so that it'll pull the latest preview package. But with these prerelease versions, if `1.1.1` already exists, `^1.1.1-feature.0` will install `1.1.1`.
> 
> One way around this is to patch bump to `1.1.2-feature.0` but then if `1.1.2` is published because someone else merges a `version packages` pr, the person working on the feature pr will start installing the new `1.1.2` or `1.2.0` instead of their preview package.
> 
> Also, another thing I noticed was there was some caching(?) issues with fetching some of these packages. Say if I have `^.0` and recently published `.1`, I should be able to install `^.0` and get `.1`. And afterwards if I publish `.2` and `.3`, installing `^.0` should get me `.3` but sometimes it will keep installing `.1` until I change my dependency to `^.2` which only _then_ it'll install `.3` and _then_ I can bump it back down to `^.0` to get `.3`.
> 
> It was at this point that we thought maybe it's good to be explicit in which preview package we want to install (instead of using carets). Also, it was my opinion that connecting preview packages to the commit sha is useful because the developer can check via `git log` and `npm view` to see if their preview package has been published. If i'm at `1.1.1-feature.8`, i might forget if `.8` was from an earlier commit or if it's the latest one.
> 
> [This](https://github.com/jbolda/covector/issues/147#issuecomment-812678832) is where I jot down my findings around prerelease versioning for covector (under the `Consuming Previews` section).

Yea, it seems that the user will need to pin regardless of what the version comes out to be. Within a specific series of versions, I think it still makes sense to keep them sequential. I think there are cases where using a `^` makes sense for ease on ongoing upgrades. It would make upgrading seamless for 95% of the cases (where publishing only happens so often). Considering that, a timestamp or an `npm view ...` would achieve that objective. I think that objective is more important than including the git sha if we have a choice. It would be great if we could somehow expose both, but I don't know that we figured that out.

I believe the issue that we were seeing was related to both dependency lists on syncing / updating with the preview version (which it will now), and with use of the `^`. Mostly the former issue if I recall correctly.

**Comment by cowboyd** [comment]
What do you guys think about pinning with a label?

In other words, `yarn add effection@pr-label`. That will resolve to the date or sha based preview version, and you won't need to think about it again unless you want to upgrade, but another `yarn add effection@pr-label` should work there since it will now resolve to the newer version.

This seems the best solution. 

1. You only have to refer to the symbolic branch
2. We don't muck with semver where we had the problem that `1.2.1-aonuhonoegfxff` was installable as a valid solution for `^1.2.1


Basically I'm ok with having less scrutable version numbers, if I never have to actually use them.

**Comment by minkimcello** [comment]
@jbolda 
> Considering that, a timestamp or an `npm view ...` would achieve that objective. I think that objective is more important than including the git sha if we have a choice. It would be great if we could somehow expose both, but I don't know that we figured that out.

Good thing covector offers both. 👍 But my opinion is that commit SHA is more helpful than the timestamp for two reasons: I often don't remember when I push a commit so the timestamp won't tell me much if I'm publishing more than one preview packages a day, and even if I do keep track of what time I push my commits, it doesn't align with when the timestamp actually happens within covector.

---

@cowboyd

The current setup should allow us to use labels. (This reminded me that I need to modify the publish command to [tag](https://github.com/jbolda/covector/pull/200)). 

With the current setup you should already be able to `yarn add pkg@label`. And if you want a newer iteration of a preview package, you will need to either update your package.json with the exact version or run `yarn remove ...` before running `yarn add pkg@label`.

**Comment by jbolda** [comment]
> The current setup should allow us to use labels. (This reminded me that I need to modify the publish command to [tag](https://github.com/jbolda/covector/pull/200)).
> 
> With the current setup you should already be able to `yarn add pkg@label`. And if you want a newer iteration of a preview package, you will need to either update your package.json with the exact version or run `yarn remove ...` before running `yarn add pkg@label`.

It kind of seems that the general mindshare is switching to npm@7. With that, we could do `npm update` on any of the label based deps, right? I don't know that I have ever tried it.

**Comment by cowboyd** [comment]
What we really need to avoid in either case with the sha is having `1.1.1-beta.fxboffiw` be a satisfactory version for `^1.1.1`

If that's the case, then I'm totally 👌🏼using the SHA, and would also prefer it to date stamps for the reasons that @minkimcello mentioned and also since it will relate groups of packages together even if their publication spans multiple minutes.

**Comment by minkimcello** [comment]
Screen Shot 2021-07-13 at 5 01 29 PM

**Comment by cowboyd** [comment]
@minkimcello does the same hold for the alphanumeric values present in a SHA?

**Comment by minkimcello** [comment]
> @minkimcello does the same hold for the alphanumeric values present in a SHA?

The tags are overwritten each publish so if we're installing with the tag, it'll always get the last published package regardless of value.

**Comment by cowboyd** [comment]
@minkimcello We had a pretty good discussion about the optimal workflow on Friday. What do we need to do to make this actionable?

**Comment by minkimcello** [comment]
> @minkimcello We had a pretty good discussion about the optimal workflow on Friday. What do we need to do to make this actionable?

With the incremental preview versions? That'll need to be worked out in covector. I'll do another dive and put together a write up sometime in the next few days. But I don't think it should be a blocker for this PR. I can come back and update the dependency of covector afterwards.

**Comment by cowboyd** [comment]
Ok, so just to be clear, what will preview packages look like for v4 prs?

**Comment by minkimcello** [comment]
@cowboyd `material-ui-interactors@4.0.0-mk-publish.1626208194558`

It's using `date` (default) but I can switch it to `sha`

**Comment by cowboyd** [comment]
Let's keep it with date.

**Review by cowboyd — APPROVED** [review]
Looks good. Ideally, we can get to cleaner pre-release version numbers in the future, but understood that it will require significant changes to covector.
